### PR TITLE
[STORM-1570] Storm SQL support for nested fields and array

### PR DESCRIPTION
**Pull request by arunmahadevan**
Added support to handle array, map and nested field lookups and user defined functions
in Storm SQL.

### CODE REVIEW

File: external/sql/storm-sql-core/src/jvm/org/apache/storm/sql/compiler/ExprCompiler.java [hunk 1]
```diff
@@ -234,14 +253,20 @@ public String translate(ExprCompiler compiler, RexCall call) {
               pw.print(String.format("else if (%2$s == null) { %1$s = null; }\n", val, arg));
             }
           }
-          String calc = printMethodCall(method.method, args);
-          pw.print(String.format("else { %1$s = %2$s; }\n", val, calc));
+          String calc = printMethodCall(method, args);
+          String valCast = compiler.javaTypeName(call);
+          pw.print(String.format("else { %1$s = (%2$s)%3$s; }\n", val, valCast, calc));
```
Comment (satishd) [inline external/sql/storm-sql-core/src/jvm/org/apache/storm/sql/compiler/ExprCompiler.java]: It seems there is no need to cast here. Tried with supported functions and UDF and it is working fine without casting. You may want to remove this as casting in generating code looks redundant.

**Comment by satishd** [comment]
Overall LGTM. +1 once the comment is addressed.

**Comment by arunmahadevan** [comment]
@haohui In some cases like `true AND null` the query was still returning results, hence I had changed `StormSqlFunctions.eq` and `StormSqlFunctions.ne` to return false if either arg was null. The actual issue was in AND_EXPR and OR_EXPR, so I fixed it and reverted the earlier change. I have also split the change into two commits to review easily. 

Currently the UDF schema definition is not exposed via SQL CREATE. Let me know if you have any inputs on how this could be done.

**Comment by haohui** [comment]
Thanks for updating the patch!

> In some cases like true AND null the query was still returning results, hence I had changed StormSqlFunctions.eq and StormSqlFunctions.ne to return false if either arg was null. 

That sounds like a bug to me. Can you please create a jira and post the query?

Posted inlined comment. I think the first commit is pretty good. Let's separate the second commit to another jira.

There are some pending tasks. I think we need to expose array / UDF in the DDL as well, my suggestion is to closely follow the syntax of Hive. I think both work can be done in separate jiras.

**Comment by arunmahadevan** [comment]
- Removed the second commit. I will raise follow up JIRAs to support UDF and expose it via DDL.
- Removed guava test dep from storm-sql-runtime (not sure why this was a concern since storm-sql-core tests already uses guava classes).
- Added test cases to test the semantics for AND and OR with nullable on both sides.

**Comment by harshach** [comment]
+1. @haohui can you check the final commit. Thanks

**Comment by haohui** [comment]
The patch looks good. There are a few nits in the latest commit, +1 once they are addressed.

Thanks for the work!

**Comment by arunmahadevan** [comment]
Addressed latest comments.

**Pull request merged**